internal/ad/ad.go

```diff
@@ -491,116 +491,155 @@ func (ad *AD) parseGPOs(ctx context.Context, gpos []gpo, objectClass ObjectClass
 			Rules: make(map[string][]entry.Entry),
 		}
 		r = append(r, gpoWithRules)
-		if err := func() error {
-			ad.downloadables[name].mu.RLock()
-			defer ad.downloadables[name].mu.RUnlock()
-			_ = ad.downloadables[name].testConcurrent
-
-			log.Debugf(ctx, "Parsing GPO %q", name)
-
-			// We need to consider the uppercase version of the name as well,
-			// which could occur in some of the default GPOs such as Default
-			// Domain Policy.
-			classes := []string{"User", "USER"}
-			if objectClass == ComputerObject {
-				classes = []string{"Machine", "MACHINE"}
+		if err = ad.parseGPO(ctx, name, url, keyFilterPrefix, objectClass, gpoWithRules); err != nil {
+			return r, err
+		}
+	}
+
+	return r, nil
+}
+
+func (ad *AD) parseGPO(ctx context.Context, name, url, keyFilterPrefix string, objectClass ObjectClass, gpoWithRules policies.GPO) error {
+	ad.downloadables[name].mu.RLock()
+	defer ad.downloadables[name].mu.RUnlock()
+	_ = ad.downloadables[name].testConcurrent
+
+	log.Debugf(ctx, "Parsing GPO %q of class %q", name, objectClass)
+
+	// We need to consider the uppercase version of the name as well,
+	// which could occur in some of the default GPOs such as Default
+	// Domain Policy.
+	classes := []string{"User", "USER"}
+	if objectClass == ComputerObject {
+		classes = []string{"Machine", "MACHINE"}
+	}
+
+	var err error
+	var f *os.File
+	for _, class := range classes {
+		var e error
+		var files []os.DirEntry
+
+		policyDir := filepath.Join(ad.sysvolCacheDir, "Policies", filepath.Base(url), class)
+		files, e = os.ReadDir(policyDir)
+
+		if e != nil {
+			if errors.Is(e, fs.ErrNotExist) {
+				log.Debugf(ctx, "Not found policy directory %q", policyDir)
+			} else {
+				log.Warningf(ctx, "Failed to read policy directory %q: %s", policyDir, e)
 			}
+		} else {
+			log.Debugf(ctx, "Success reading policy directory %q: found %d files", policyDir, len(files))
+		}
+
+		foundPolicy := false
 
-			var err error
-			var f *os.File
-			for _, class := range classes {
-				var e error
-				f, e = os.Open(filepath.Join(ad.sysvolCacheDir, "Policies", filepath.Base(url), class, "Registry.pol"))
-
-				// We only care about the first error which is caused by opening
-				// the capitalized version of the class, instead of the
-				// uppercase version which is less common and more of an edge case.
-				if e != nil && err == nil {
-					err = e
-				} else if e == nil {
-					err = nil
-					break
+		//The filesystem might be case-sensitive, and if the registry policy is created as registry.pol
+		//it will not be found.
+		for _, file := range files {
+			if strings.EqualFold(file.Name(), "Registry.pol") {
+				policyPath := filepath.Join(policyDir, file.Name())
+				log.Debugf(ctx, "Found registry policy file %q", policyPath)
+
+				f, e = os.Open(policyPath)
+
+				if e != nil {
```

Comment:
See my previous comment on early return on error. Also here, as we can’t have multiple "classes" where we could read the directory, I guess there are 2 kinds of errors:
- the file doesn’t exist -> ok, do next
- any other kind of return -> early return, the policy file has an issue (do not ignore it silently and try to continue).

Something like what we have below:
```go
if errors.Is(err, fs.ErrNotExist) {
	log.Debugf(ctx, "Policy %q doesn't have any policy for class %q %s", name, objectClass, e)
	return nil
} else if e != nil {
	return e
}
```

Reply:
Simplified the code to fail fast.